shared/src/baseline_solver.rs

```diff
@@ -20,78 +20,145 @@ pub trait BaselineSolvable {
     fn get_spot_price(&self, base_token: H160, quote_token: H160) -> Option<BigRational>;
 }
 
-// Given a path and sell amount (first token of the path) estimates the buy amount (last token of the path).
+pub struct Estimate<'a, V, L> {
+    // The result amount of the estimate
+    pub value: V,
+    // The liquidity path used to derive at that estimate
+    pub path: Vec<&'a L>,
+}
+
+// Given a path and sell amount (first token of the path) estimates the buy amount (last token of the path) and
+// the path of liquidity that yields this result
 // Returns None if the path is invalid or pool information doesn't exist.
-pub fn estimate_buy_amount(
+pub fn estimate_buy_amount<'a, L: BaselineSolvable>(
     sell_amount: U256,
     path: &[H160],
-    liquidity: &HashMap<TokenPair, impl BaselineSolvable>,
-) -> Option<U256> {
+    liquidity: &'a HashMap<TokenPair, Vec<L>>,
+) -> Option<Estimate<'a, U256, L>> {
     let sell_token = path.first()?;
     path.iter()
         .skip(1)
-        .fold(Some((sell_amount, *sell_token)), |previous, current| {
-            let previous = match previous {
-                Some(previous) => previous,
-                None => return None,
-            };
-
-            match liquidity.get(&TokenPair::new(*current, previous.1)?) {
-                Some(liquidity) => liquidity
-                    .get_amount_out(*current, previous.0, previous.1)
-                    .map(|amount| (amount, *current)),
-                None => None,
-            }
+        .fold(
+            Some((sell_amount, *sell_token, Vec::new())),
+            |previous, current| {
+                let previous = match previous {
+                    Some(previous) => previous,
+                    None => return None,
+                };
+
+                match liquidity.get(&TokenPair::new(*current, previous.1)?) {
+                    Some(liquidity) => liquidity
+                        .iter()
+                        .max_by_key(|item| item.get_amount_out(*current, previous.0, previous.1))
+                        .and_then(|best| {
+                            best.get_amount_out(*current, previous.0, previous.1)
+                                .map(|amount| {
+                                    (
+                                        amount,
+                                        *current,
+                                        previous
+                                            .2
+                                            .into_iter()
+                                            .chain(std::iter::once(best))
+                                            .collect(),
+                                    )
+                                })
+                        }),
+                    None => None,
+                }
+            },
+        )
```

Comment:
I have simplified the part in `fold`. You should be able to do something similar for the sell amount case.
```suggestion
        .fold(
            Some((sell_amount, *sell_token, Vec::new())),
            |previous, current| {
                let (amount, previous, mut path) = previous?;
                let best = liquidity
                    .get(&TokenPair::new(*current, previous)?)?
                    .iter()
                    .map(|l| (l, l.get_amount_out(*current, amount, previous)))
                    .max_by(|a, b| a.1.cmp(&b.1))?;
                path.push(best.0);
                Some((best.1?, *current, path))
            },
        )
```